python/instrumentation/openinference-instrumentation-llama-index/tests/openinference/instrumentation/llama_index/test_callback.py

```diff
@@ -0,0 +1,354 @@
+    assert retrieve_attributes.pop(OPENINFERENCE_SPAN_KIND, None) == RETRIEVER.value
+    assert retrieve_attributes.pop(INPUT_VALUE, None) == question
+    retrieve_attributes.pop(f"{RETRIEVAL_DOCUMENTS}.0.{DOCUMENT_ID}", None)
+    retrieve_attributes.pop(f"{RETRIEVAL_DOCUMENTS}.1.{DOCUMENT_ID}", None)
+    assert (
+        retrieve_attributes.pop(f"{RETRIEVAL_DOCUMENTS}.0.{DOCUMENT_CONTENT}", None)
+        == nodes[0].text
+    )
+    assert (
+        retrieve_attributes.pop(f"{RETRIEVAL_DOCUMENTS}.0.{DOCUMENT_METADATA}.category", None)
+        == nodes[0].metadata["category"]
+    )
+    assert (
+        retrieve_attributes.pop(f"{RETRIEVAL_DOCUMENTS}.1.{DOCUMENT_CONTENT}", None)
+        == nodes[1].text
+    )
+    assert retrieve_attributes == {}
+
+    if status_code == 200:
+        # FIXME: LlamaIndex doesn't currently capture the LLM span when status_code == 400
+        # For example, if an exception is raised by the LLM at the following location,
+        # `on_event_end` never gets called.
+        # https://github.com/run-llama/llama_index/blob/dcef41ee67925cccf1ee7bb2dd386bcf0564ba29/llama_index/llms/base.py#L100 # noqa E501
+        assert (llm_span := spans_by_name.pop("llm", None)) is not None
+        assert llm_span.parent is not None
+        assert llm_span.parent.span_id == synthesize_span.context.span_id
+        assert llm_span.context.trace_id == synthesize_span.context.trace_id
+        llm_attributes = dict(llm_span.attributes or {})
+        assert llm_attributes.pop(OPENINFERENCE_SPAN_KIND, None) == LLM.value
+        assert llm_attributes.pop(LLM_MODEL_NAME, None) is not None
+        assert llm_attributes.pop(LLM_INVOCATION_PARAMETERS, None) is not None
+        assert llm_attributes.pop(LLM_PROMPT_TEMPLATE, None) is not None
+        assert llm_attributes.pop(f"{LLM_PROMPT_TEMPLATE_VARIABLES}.context_str", None) is not None
+        assert llm_attributes.pop(f"{LLM_PROMPT_TEMPLATE_VARIABLES}.query_str", None) == question
+        assert llm_attributes.pop(f"{LLM_INPUT_MESSAGES}.0.{MESSAGE_ROLE}", None) is not None
+        assert llm_attributes.pop(f"{LLM_INPUT_MESSAGES}.0.{MESSAGE_CONTENT}", None) is not None
+        assert llm_attributes.pop(f"{LLM_INPUT_MESSAGES}.1.{MESSAGE_ROLE}", None) is not None
+        assert llm_attributes.pop(f"{LLM_INPUT_MESSAGES}.1.{MESSAGE_CONTENT}", None) is not None
+        if status_code == 200:
+            assert llm_span.status.status_code == trace_api.StatusCode.OK
+            assert llm_attributes.pop(OUTPUT_VALUE, None) == answer
+            if not is_stream:
+                # FIXME: currently we can't capture messages when streaming
+                assert (
+                    llm_attributes.pop(f"{LLM_OUTPUT_MESSAGES}.0.{MESSAGE_ROLE}", None)
+                    == "assistant"
+                )
+                assert (
+                    llm_attributes.pop(f"{LLM_OUTPUT_MESSAGES}.0.{MESSAGE_CONTENT}", None) == answer
+                )
```

Comment:
This conditional is unnecessary.

Reply:
you're right. This was written before i realize the whole thing doesn't work when it's 400.